File: core/src/test/kotlin/com/example/toyracers/collision/CollisionEdgeCaseTest.kt
package com.example.toyracers.collision

import com.example.toyracers.car.CarState
import com.example.toyracers.track.Track
import com.example.toyracers.track.TrackCircle
import com.example.toyracers.track.TrackLoader
import com.example.toyracers.track.TrackPoint
import com.example.toyracers.track.TrackPolygon
import com.example.toyracers.track.TrackRectangle
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class CollisionEdgeCaseTest {
    private val collisionSystem = CollisionSystem()
    private val baseTrack = TrackLoader().load()

    @Test
    fun `inside rectangle resolves through its nearest side`() {
        val obstacle = TrackRectangle(20f, 20f, 4f, 4f)
        val cases =
            listOf(
                RectangleCase(20.5f, 22f, 19f, 22f),
                RectangleCase(23.5f, 22f, 25f, 22f),
                RectangleCase(22f, 20.5f, 22f, 19f),
                RectangleCase(22f, 23.5f, 22f, 25f),
            )

        cases.forEach { case ->
            val state = CarState(x = case.x, y = case.y)

            val result = collisionSystem.resolveTrackCollision(state, radius = 1f, rectangleTrack(obstacle))

            assertTrue(result.collided)
            assertEquals(CollisionType.TRACK_OBJECT, result.contacts.single().type)
            assertEquals(case.expectedX, state.x, TOLERANCE)
            assertEquals(case.expectedY, state.y, TOLERANCE)
        }
    }

    @Test
    fun `rectangle corner contact uses diagonal normal`() {
        val state = CarState(x = 19.5f, y = 19.5f, velocityX = 2f, velocityY = 2f)

        val result =
            collisionSystem.resolveTrackCollision(
                state,
                radius = 1f,
                rectangleTrack(TrackRectangle(20f, 20f, 4f, 4f)),
            )

        val contact = result.contacts.single()
        assertTrue(result.collided)
        assertTrue(contact.normalX < 0f)
        assertTrue(contact.normalY < 0f)
        assertTrue(state.x < 19.5f)
        assertTrue(state.y < 19.5f)
    }

    @Test
    fun `circle collision handles center overlap and exact tangent`() {
        val circle = TrackCircle(TrackPoint(30f, 30f), radius = 2f)
        val centered = CarState(x = 30f, y = 30f)

        val centeredResult = collisionSystem.resolveTrackCollision(centered, radius = 1f, shapeTrack(circle))
        val tangent = CarState(x = 33f, y = 30f)
        val tangentResult = collisionSystem.resolveTrackCollision(tangent, radius = 1f, shapeTrack(circle))

        assertTrue(centeredResult.collided)
        assertEquals(33f, centered.x, TOLERANCE)
        assertFalse(tangentResult.collided)
        assertEquals(33f, tangent.x, TOLERANCE)
    }

    @Test
    fun `polygon collision resolves interior for either winding direction`() {
        val counterClockwise = squareVertices()
        val clockwise = counterClockwise.reversed()

        listOf(counterClockwise, clockwise).forEach { vertices ->
            val state = CarState(x = 42f, y = 41f)

            val result =
                collisionSystem.resolveTrackCollision(
                    state,
                    radius = 1f,
                    shapeTrack(TrackPolygon(vertices)),
                )

            assertTrue(result.collided)
            assertEquals(CollisionType.TRACK_OBJECT, result.contacts.single().type)
            assertEquals(39f, state.y, TOLERANCE)
        }
    }

    @Test
    fun `world boundary resolves right and top edges and preserves inward motion`() {
        val state = CarState(x = baseTrack.worldBounds.maxX + 2f, y = baseTrack.worldBounds.maxY + 2f)
        state.velocityX = -3f
        state.velocityY = -4f

        val result =
            collisionSystem.resolveTrackCollision(
                state,
                radius = 1f,
                baseTrack.copy(collisionShapes = emptyList()),
            )

        assertEquals(baseTrack.worldBounds.maxX - 1f, state.x, TOLERANCE)
        assertEquals(baseTrack.worldBounds.maxY - 1f, state.y, TOLERANCE)
        assertEquals(-3f, state.velocityX, TOLERANCE)
        assertEquals(-4f, state.velocityY, TOLERANCE)
        assertEquals(2, result.contacts.size)
    }

    @Test
    fun `car collision resolves coincident centers and ignores tangent cars`() {
        val first = CarState(x = 50f, y = 50f)
        val second = CarState(x = 50f, y = 50f)

        val overlap = collisionSystem.resolveCarCollision(first, 1f, second, 1f)
        val tangentFirst = CarState(x = 60f, y = 60f)
        val tangentSecond = CarState(x = 62f, y = 60f)
        val tangent = collisionSystem.resolveCarCollision(tangentFirst, 1f, tangentSecond, 1f)

        assertTrue(overlap.collided)
        assertEquals(49f, first.x, TOLERANCE)
        assertEquals(51f, second.x, TOLERANCE)
        assertFalse(tangent.collided)
        assertEquals(60f, tangentFirst.x, TOLERANCE)
        assertEquals(62f, tangentSecond.x, TOLERANCE)
    }

    private fun rectangleTrack(obstacle: TrackRectangle): Track =
        baseTrack.copy(innerObstacles = listOf(obstacle), collisionShapes = emptyList())

    private fun shapeTrack(shape: com.example.toyracers.track.TrackCollisionShape): Track =
        baseTrack.copy(innerObstacles = emptyList(), collisionShapes = listOf(shape))

    private fun squareVertices(): List<TrackPoint> =
        listOf(
            TrackPoint(40f, 40f),
            TrackPoint(44f, 40f),
            TrackPoint(44f, 44f),
            TrackPoint(40f, 44f),
        )

    private data class RectangleCase(
        val x: Float,
        val y: Float,
        val expectedX: Float,
        val expectedY: Float,
    )

    private companion object {
        const val TOLERANCE = 0.001f
    }
}

File: dart/test/collision_edge_case_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:toy_racers/simulation.dart';
import 'package:toy_racers/simulation/collision/collision_geometry.dart';

import 'collision_test_support.dart';

void main() {
  final collisionSystem = CollisionSystem();
  final unitCircleConfig = CarConfig(
    collisionRadius: 1,
    collisionLongitudinalOffset: 0,
    width: 2,
    length: 3,
  );

  group('CollisionSystem geometry and impact state', () {
    test(
      'keeps double-precision hypot intermediates at the binary32 boundary',
      () {
        expect(
          CollisionGeometry.hypot(132.68743896484375, -117.60503387451172),
          177.30453491210938,
        );
        expect(CollisionGeometry.minimumDistance, 0.00009999999747378752);

        final state = CarState(velocityX: 0.0001);
        CollisionGeometry.updateLongitudinalSpeed(state);

        expect(state.velocityX, Float32.narrow(0.0001));
        expect(state.longitudinalSpeed, Float32.narrow(0.0001));
      },
    );

    test('retains the legacy max impact speed constructor', () {
      final result = CollisionResult(maxImpactSpeed: 2.5);

      expect(result.contacts, isEmpty);
      expect(result.maxImpactSpeed, Float32.narrow(2.5));
    });

    test(
      'world edge removes outward velocity and retains tangential speed',
      () {
        final state = CarState(x: -1, y: 6, velocityX: -10, velocityY: 4);

        final result = resolveTrackCollisionForTest(
          collisionSystem,
          state,
          unitCircleConfig,
        );

        expect(result.collided, isTrue);
        expect(result.contacts.single.type, CollisionType.worldBoundary);
        expect(state.x, closeTo(unitCircleConfig.collisionRadius, _tolerance));
        expect(state.velocityX, 0);
        expect(state.velocityY, closeTo(2.6, _tolerance));
        expect(result.contacts.single.impactSpeed, 10);
        expect(result.maxImpactSpeed, 10);
      },
    );

    test(
      'separating boundary contact keeps inward velocity and has no impact',
      () {
        final state = CarState(x: 0.2, y: 6, velocityX: 3);

        final result = resolveTrackCollisionForTest(
          collisionSystem,
          state,
          unitCircleConfig,
        );

        expect(result.collided, isTrue);
        expect(state.velocityX, 3);
        expect(result.contacts.single.impactSpeed, 0);
      },
    );

    test('inside rectangle resolves by nearest side in fixed tie order', () {
      final obstacle = TrackRectangle(20, 20, 4, 4);
      final cases =
          <({double x, double y, double expectedX, double expectedY})>[
            (x: 20.5, y: 22, expectedX: 19, expectedY: 22),
            (x: 23.5, y: 22, expectedX: 25, expectedY: 22),
            (x: 22, y: 20.5, expectedX: 22, expectedY: 19),
            (x: 22, y: 23.5, expectedX: 22, expectedY: 25),
          ];

      for (final scenario in cases) {
        final state = CarState(x: scenario.x, y: scenario.y);
        final result = resolveTrackCollisionForTest(
          collisionSystem,
          state,
          unitCircleConfig,
          innerObstacles: <TrackRectangle>[obstacle],
        );
        expect(result.contacts, hasLength(1), reason: '$scenario');
        expect(result.contacts.single.type, CollisionType.trackObject);
        expect(state.x, closeTo(scenario.expectedX, _tolerance));
        expect(state.y, closeTo(scenario.expectedY, _tolerance));
      }
    });

    test('circle contact handles center overlap but not exact tangent', () {
      final circle = TrackCircle(center: TrackPoint(30, 30), radius: 2);
      final centered = CarState(x: 30, y: 30);
      final tangent = CarState(x: 33, y: 30);

      final centeredResult = resolveTrackCollisionForTest(
        collisionSystem,
        centered,
        unitCircleConfig,
        collisionShapes: <TrackCollisionShape>[circle],
      );
      final tangentResult = resolveTrackCollisionForTest(
        collisionSystem,
        tangent,
        unitCircleConfig,
        collisionShapes: <TrackCollisionShape>[circle],
      );

      expect(centeredResult.collided, isTrue);
      expect(centered.x, 33);
      expect(tangentResult.collided, isFalse);
      expect(tangent.x, 33);
    });

    test('polygon collision resolves interior for both winding directions', () {
      final counterClockwise = <TrackPoint>[
        TrackPoint(40, 40),
        TrackPoint(44, 40),
        TrackPoint(44, 44),
        TrackPoint(40, 44),
      ];

      for (final vertices in <List<TrackPoint>>[
        counterClockwise,
        counterClockwise.reversed.toList(),
      ]) {
        final state = CarState(x: 42, y: 41);
        final result = resolveTrackCollisionForTest(
          collisionSystem,
          state,
          unitCircleConfig,
          collisionShapes: <TrackCollisionShape>[TrackPolygon(vertices)],
        );
        expect(result.contacts.single.type, CollisionType.trackObject);
        expect(state.y, closeTo(39, _tolerance));
      }
    });

    test('coincident cars separate along the fixed fallback normal', () {
      final first = CarState(x: 50, y: 50);
      final second = CarState(x: 50, y: 50);

      final result = collisionSystem.resolveCarCollision(
        firstState: first,
        firstConfig: unitCircleConfig,
        secondState: second,
        secondConfig: unitCircleConfig,
      );

      expect(result.collided, isTrue);
      expect(first.x, 49);
      expect(second.x, 51);
      expect(result.contacts.single.normalX, -1);
      expect(result.contacts.single.impactSpeed, 0);
    });

    test('tangent cars do not report a contact', () {
      final first = CarState(x: 60, y: 60);
      final second = CarState(x: 62, y: 60);

      final result = collisionSystem.resolveCarCollision(
        firstState: first,
        firstConfig: unitCircleConfig,
        secondState: second,
        secondConfig: unitCircleConfig,
      );

      expect(result.collided, isFalse);
      expect(first.x, 60);
      expect(second.x, 62);
    });

    test('configuration and resolution reject invalid collision settings', () {
      expect(
        () => CollisionConfig(wallSpeedRetention: -0.1),
        throwsArgumentError,
      );
      expect(() => CollisionConfig(carRestitution: 1.1), throwsArgumentError);
      expect(() => CollisionConfig(maxCarImpulse: -1), throwsArgumentError);
      expect(
        () => CarConfig(collisionRadius: 100, width: 200, length: 202),
        returnsNormally,
      );
      final oversizedConfig = CarConfig(
        collisionRadius: 100,
        width: 200,
        length: 202,
      );
      final state = CarState(x: 4, y: 6, rotationDegrees: 45);

      expect(
        () => resolveTrackCollisionForTest(
          collisionSystem,
          state,
          oversizedConfig,
        ),
        throwsArgumentError,
      );
      expect(state.x, 4);
      expect(state.y, 6);
    });
  });
}

const double _tolerance = 0.001;
